### PR TITLE
Jguer/ /add pre receive hooks

### PR DESCRIPTION
**Pull request by Jguer**
commit 1:
- refactor `--mode` into sub commands (`--mode` will still work for now but it's hidden)
- split the old scan if train into separate commands
- CLI changes config object after configs are loaded.

commit 2:
- Adds commit-range for pre-receive hooks (also useful for testing and scanning defined parts of a repo (like a branch))

commit 3 and 4:
- Documentation and fixes

Some type hinting corrections sprinkled on top

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -276,11 +331,23 @@ If you already have a pre-commit executable file and you want to use gg-shield,
 all you need to do is to add this line in the file:
 
 ```shell
-ggshield scan --mode pre-commit
+ggshield scan pre-commit
 ```
 
 Do not forget to add your [GitGuardian API Key](https://dashboard.gitguardian.com/api/v1/auth/user/github_login/authorize?utm_source=github&utm_medium=gg_shield&utm_campaign=shield1) to the `GITGUARDIAN_API_KEY` environment variable of your project or development environment.
 
+# Pre-receive
```
Comment (ericfourrier) [inline README.md:339]:
```suggestion
# Git pre-receive hooks
```

Comment (Jguer) [reply README.md]: Fixed in following commit, index also required change

File: ggshield/cmd.py [hunk 2]
```diff
@@ -30,152 +31,124 @@
...
+    "The env var GITGUARDIAN_EXIT_ZERO can also be used to set this option.",
 )
 @click.option(
     "--all-policies",
     is_flag=True,
     default=None,
     help="Present fails of all policies (Filenames, FileExtensions, Secret Detection)."
-    "By default, only Secret Detection is shown"
-    "",
+    " By default, only Secret Detection is shown.",
```
Comment (eugenenelou) [inline ggshield/cmd.py:54]:
```suggestion
    "By default, only Secret Detection is shown.",
```

File: ggshield/dev_scan.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+@click.command()
+@click.argument("repository", nargs=1, type=click.STRING, required=True)
+@click.pass_context
+def repo_cmd(ctx: click.Context, repository: str) -> int:
+    """
+    clone and scan a REPOSITORY.
+
+    REPOSITORY is the clone URI of the repository to scan
+    example:
```
Comment (eugenenelou) [inline ggshield/dev_scan.py:36]: :question: Did you want to add an example here?

File: ggshield/message.py [hunk 4]
```diff
@@ -303,3 +303,27 @@ def get_lines_to_display(
                 )
 
     return lines_to_display
+
+
+def process_results(results: List[Result], show_secrets: bool, verbose: bool):
```
Comment (eugenenelou) [inline ggshield/message.py:308]:
```suggestion
def process_results(results: List[Result], show_secrets: bool, verbose: bool) -> int:
```

File: ggshield/git_shell.py [hunk 5]
```diff
@@ -58,7 +62,10 @@ def get_list_commit_SHA(commit_range: str) -> List[str]:
     :param commit_range: A range of commits (ORIGIN...HEAD)
     """
 
-    return shell([GIT_PATH, "rev-list", "--reverse", commit_range])
+    commit_list = shell([GIT_PATH, "rev-list", "--reverse", *commit_range.split()])
+    if "" in commit_list:
+        commit_list.remove("")
```
Comment (eugenenelou) [inline ggshield/git_shell.py:67]: :question: `remove` only removes one occurrence, the empty string can only appear once? (I think I remember it's a trailing new line that cause it)

Comment (Jguer) [reply ggshield/git_shell.py:67]: Only happens when the `git rev-list` returns an empty rev-list but no error example `git rev-list HEAD..HEAD`

**Pull request merged**